lib/matplotlib/__init__.py

```diff
@@ -1515,6 +1515,43 @@ def test(verbosity=1):
 
 test.__test__ = False  # nose: this function is not a test
 
+
+def _replacer(data, key):
+    # if key isn't a string don't bother
+    if not isinstance(key, six.string_types):
+        return key
+    # try to use __getitem__
+    try:
+        return data[key]
+    # key does not exist, silently fall back to key
+    except KeyError:
+        return key
+
+
+def unpack_labeled_data(func):
+    """
+    A decorator to add a 'data' kwarg to any a function.  The signature
+    of the input function must be ::
+
+       def foo(ax, *args, **kwargs)
+
+    so this is suitable for use with Axes methods.
+    """
+    @functools.wraps(func)
+    def inner(ax, *args, **kwargs):
+        data = kwargs.pop('data', None)
+        if data is not None:
+            if rcParams['unpack_labeled']:
+                args = tuple(_replacer(data, a) for a in args)
+                kwargs = dict((k, _replacer(data, v))
+                              for k, v in six.iteritems(kwargs))
```

Comment:
erk, didn't know that :-(

Not sure if that helps, but:
```
def func(x, y=1):
    print("x: %s, y: %s" % (x,y))
inspect.getargspec(func)
ArgSpec(args=['x', 'y'], varargs=None, keywords=None, defaults=(1,))
```

0. only pass in the names of the args in the decorator as `replace_names`
1. in the decorator: cache the list of arg names via inspect: `cached_names`
2 in the wrapper: 
  1. for each arg, use the `pos` to get the name (`cached_names[pos]`) and use that name in the replacement if it is in `replace_names`
  2. process all kwargs like before if they are in `replace_names`

Unfortunately this won't work for `plot_func(*arg, **kwarg)` :-(